### PR TITLE
feat: Add parquet upload

### LINKED ISSUE #14020: 'Upload parquet' option
**Is your feature request related to a problem? Please describe.**
In my data mining department we read data from parquets, something that's quite common and widespread. However, Superset doesn't give the option to directly upload a parquet file, just csv or excel.
**Describe the solution you'd like**
I would like it to be an 'upload parquet' option, alongside the existing ones.
**Describe alternatives you've considered**
I am already using Drill with Superset since Drill supports many kind of files. It would be perfect were it to be an option like when adding the database and using Drill, of using SQLAlchemy to connect with more databases that the default ones.

### PR DESCRIPTION
**Pull request by exemplary-citizen**
### SUMMARY
Allow CSV upload form to accept parquet file. Went in this direction so as not to exacerbate what was brought up in #13834 by adding a new form specifically for parquet files. I believe small modifications can be made to this PR to accommodate `feather` and `orc` files.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
image

### TEST PLAN
Added a test similar to the ones already in `csv_upload_tests.py`

### ADDITIONAL INFORMATION
Fixes #14020 
- [x] Has associated issue: #14020 
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by exemplary-citizen** [comment]
@john-bodley @villebro just wanted to follow up on this in case it got lost in the shuffle. How should we proceed on this PR?

**Comment by villebro** [comment]
@exemplary-citizen sorry for dropping the ball on this - I'll have this reviewed within the next 24 hours

File: superset/views/database/views.py [hunk 1]
```diff
@@ -128,6 +128,32 @@ def form_get(self, form: CsvToDatabaseForm) -> None:
...
+            kwargs = {
+                "chunksize": 1000,
+                "encoding": "utf-8",
+                "header": form.header.data if form.header.data else 0,
+                "index_col": form.index_col.data,
+                "infer_datetime_format": form.infer_datetime_format.data,
+                "iterator": True,
+                "keep_default_na": not form.null_values.data,
+                "mangle_dupe_cols": form.mangle_dupe_cols.data,
+                "usecols": form.usecols.data,
```
Comment (robdiciuccio) [inline superset/views/database/views.py:148]: This appears to change the behavior of the existing CSV upload functionality by specifying columns. Can you add some tests around this?

File: superset/views/database/forms.py [hunk 2]
```diff
@@ -163,6 +165,15 @@ def at_least_one_schema_is_allowed(database: Database) -> bool:
...
     )
+    usecols = JsonListField(
+        _("Use Columns"),
+        default=None,
+        description=_(
+            "Json list of the column names that should be read. "
+            "If not None, only these columns will be read from the file."
+        ),
+        validators=[Optional()],
+    )
```
Comment (robdiciuccio) [inline superset/views/database/forms.py:176]: Can you provide a screenshot of the updated form UI?

Comment (exemplary-citizen) [reply superset/views/database/forms.py:176]: Added a screenshot to the summary above

File: superset/views/database/views.py (hunk 1 above)
Comment (exemplary-citizen) [reply superset/views/database/views.py:148]: added a scenario to `test_import_csv` that tests uploading a CSV with specific columns

**Review by villebro — COMMENTED** [review]
A few comments:
- While it's convenient to add this functionality to the CSV upload form, I feel this should be in a form of its own, as the majority of the current fields are specific to CSV only (IIUC only `usecols` is needed for Parquet upload)
- If the CSV upload form will also handle Parquet, the title needs to be updated to reflect this. However, I'd personally prefer moving this into a form of its own.
- it would be nice if the form could handle directories/zip files, as it's fairly common to have partitioned data that is split up into multiple Parquet files. As pandas also supports uploading from a directory path, this would be a great feature to avoid having to manually append upload each file.

**Comment by exemplary-citizen** [comment]
Yeah I agree that this probably belongs in a separate form. Just went in this direction because creating a new form would mean that we'd effectively be abandoning #13834. I'll go ahead and get started working on a new form for `parquet/directory/zip` unless @junlincc @srinify @Steejay have any ideas on how to better consolidate these forms. I believe `wtforms` has a `MultipleFileField` class that should allow us to upload entire directories. We can also introduce this change to the `CsvToDatabaseForm` so users can upload multiple CSVs at a time.

**Comment by exemplary-citizen** [comment]
Here is what the new form looks like:
Screen Shot 2021-06-06 at 7 50 02 PM

File: superset/views/database/forms.py [hunk 3]
```diff
@@ -163,6 +169,15 @@ def at_least_one_schema_is_allowed(database: Database) -> bool:
...
```
Comment (exemplary-citizen) [inline superset/views/database/forms.py:180]: @villebro do you still want to keep the `Use Columns` feature in the csv form? if so, should we keep it in this PR or move it to a separate one?

**Comment by exemplary-citizen** [comment]
Not sure why CI threw a 502 error. @villebro can you enable CI again?

**Comment by exemplary-citizen** [comment]
Looks like the failing presto postgres test was fixed in #15163. CI should come back green now.

**Comment by arita37** [comment]
This is a great PR

**Comment by exemplary-citizen** [comment]
@villebro do you have any lingering doubts? if not, we should be good to merge once you enable CI and it comes back green.

**Comment by villebro** [comment]
@exemplary-citizen thanks for the ping! I restarted CI, I'll re-review today 👍

File: superset/templates/superset/form_view/columnar_to_database_view/edit.html [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    update_schemas_allowed_for_columnar_upload(db.val());
+    db.change(function(){
+        update_schemas_allowed_for_columnar_upload(db.val());
+    });
+
+    function update_schemas_allowed_for_columnar_upload(db_id) {
+        $.ajax({
+          method: "GET",
+          url: "/superset/schemas_access_for_columnar_upload",
```
Comment (villebro) [inline superset/templates/superset/form_view/columnar_to_database_view/edit.html:39]: This endpoint is called `schemas_access_for_csv_upload` and will cause a 404 (this same error happens in the Excel upload form) - either use that or rename the original to `schemas_access_for_file_upload`.

File: superset/views/database/forms.py [hunk 5]
```diff
@@ -402,3 +417,130 @@ def at_least_one_schema_is_allowed(database: Database) -> bool:
             'Use [""] for empty string.'
         ),
     )
+
+
+class ColumnarToDatabaseForm(DynamicForm):
+    # pylint: disable=E0211
+    def columnar_allowed_dbs() -> List[Database]:  # type: ignore
```
Comment (villebro) [inline superset/views/database/forms.py:424]: I wonder why this wasn't `@staticmethod`?? Can we update that here and in the other classes

**Review by villebro — CHANGES_REQUESTED** [review]
Great work! Just a few (hopefully) last comments - I'll prioritize seeing this through, so as soon as you can make the necessary updates let me know!

**Comment by villebro** [comment]
@exemplary-citizen btw, it appears one of the csv upload integration tests is failing

File: superset/views/database/forms.py [hunk 6]
```diff
@@ -402,3 +417,130 @@ def at_least_one_schema_is_allowed(database: Database) -> bool:
...
+        columnar_enabled_dbs = (
+            db.session.query(Database).filter_by(allow_csv_upload=True).all()
+        )
+        return [
+            columnar_enabled_db
+            for columnar_enabled_db in columnar_enabled_dbs
+            if ColumnarToDatabaseForm.at_least_one_schema_is_allowed(
+                columnar_enabled_db
+            )
+        ]
```
[the remark this answers is not in the document]
Comment (exemplary-citizen) [reply superset/views/database/forms.py:435]: I just opened https://github.com/apache/superset/issues/16046. Feel free to assign it to me. I'll try to open a PR for it sometime in the next week

**Comment by exemplary-citizen** [comment]
I think the failure might be coming from https://github.com/apache/superset/pull/15909. Rebasing should fix it

**Comment by arita37** [comment]
It will be a very very useful features
for many people.

File: superset/views/database/forms.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (villebro) [reply superset/views/database/forms.py:435]: Thanks - I assigned to you (will keep myself as assignee for review)

File: superset/views/database/forms.py (hunk 5 above)
Comment (exemplary-citizen) [reply superset/views/database/forms.py:424]: Adding that decorator will throw `TypeError: 'staticmethod' object is not callable`

**Comment by exemplary-citizen** [comment]
@villebro can you restart CI?

**Comment by villebro** [comment]
@exemplary-citizen sorry to bother you again, but we've recently updated the version of `pylint` which appears to be causing some new linting issues. Do you want to fix this or is it ok if I jump in and resolve the remaining ones?

**Review by villebro — APPROVED** [review]
LGTM! While testing I found some edge cases that caused trouble, but those can be improved upon later (I'll try to open up a PR for some of them; will tag you for a review when I do).

**Comment by exemplary-citizen** [comment]
@villebro thx for pushing this through. yeah do tag me, i'll be sure to review. in the meantime i'll get started on #16046

**Comment by villebro** [comment]
> @villebro thx for pushing this through. yeah do tag me, i'll be sure to review. in the meantime i'll get started on #16046

That sounds great @exemplary-citizen! Oh, and I forgot; thanks so much for your patience with the review process!

**Comment by LeoDiep** [comment]
> > @villebro thx for pushing this through. yeah do tag me, i'll be sure to review. in the meantime i'll get started on #16046
> 
> That sounds great @exemplary-citizen! Oh, and I forgot; thanks so much for your patience with the review process!

Hi @villebro and @exemplary-citizen , I am using Superset v2.1.0  docker compose and I couldn't upload a parquet file to Superset. Is this request get deprecated in new version? 
image

**Comment by LeoDiep** [comment]
> > > @villebro thx for pushing this through. yeah do tag me, i'll be sure to review. in the meantime i'll get started on #16046
> > 
> > 
> > That sounds great @exemplary-citizen! Oh, and I forgot; thanks so much for your patience with the review process!
> 
> Hi @villebro and @exemplary-citizen , I am using Superset v2.1.0 docker compose and I couldn't upload a parquet file to Superset. Is this request get deprecated in new version? image

I am sorry for the question above, I see the  `Upload columnar file to database` now.
I am trying to programatic upload file from S3 to Superset db because the performance of slice query to Athena via pyAthena are so slow compare to postgresql query directly to superset db. 
As I can see Athena doesn't  receive 25 requests at the same time (my dashboard has 25 charts in 1 page) but doing it one by one and send the result back to Superset. I have tried domain sharding, increase gunicorn workers + threads but my Superset on docker compose didn't improve performance.

Is there a way to programatically import parquet files to superset db? Thanks

**Pull request merged**